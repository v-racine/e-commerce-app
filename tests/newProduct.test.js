const { Config } = require('../src/config/config');

Config.Get(process.env);

const AppFactory = require('../src/app');
const request = require('supertest');

describe('create new product', () => {
  const mockProductsRepo = {};

  const app = AppFactory({
    productsRepo: mockProductsRepo,
  });

  describe('when: the user passes in something that is not a valid product title', () => {
    let rsp;

    beforeEach(async () => {
      rsp = await request(app).post('/admin/products/new').send({ title: 'a', price: '10' });
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    test("then: we return 'Must be between 5 and 40 characters' html", async () => {
      const status = rsp.status;
      expect(status).toBe(200);

      const text = rsp.text;
      expect(text).toMatchSnapshot();
    });
  });

  describe('when: the user passes in something that is not a float number', () => {
    let rsp;

    beforeEach(async () => {
      rsp = await request(app)
        .post('/admin/products/new')
        .send({ title: 'sweater', price: '   test   ' });
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    test("then: we return 'Invalid value'", () => {
      const status = rsp.status;
      expect(status).toBe(200);

      const text = rsp.text;
      expect(text).toMatchSnapshot();
    });
  });

  describe('when: the user passes in something that is not a product and no price', () => {
    let rsp;

    beforeEach(async () => {
      rsp = await request(app)
        .post('/admin/products/new')
        .send({ title: 'a', price: '   test   ' });
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    test('then: we return both error messages', () => {
      const status = rsp.status;
      expect(status).toBe(200);

      const text = rsp.text;
      expect(text).toMatchSnapshot();
    });
  });

  describe('when: the db already contains that product', () => {
    let rsp;

    beforeEach(async () => {
      mockProductsRepo.getOneBy = jest.fn().mockImplementation(() => {
        return new Promise((resolve) => {
          resolve({});
        });
      });

      rsp = await request(app).post('/admin/products/new').send({ title: 'testing', price: '1' });
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    test("then: we return 'Product already exists'", () => {
      const status = rsp.status;
      expect(status).toBe(200);

      const text = rsp.text;
      expect(text).toMatchSnapshot();

      expect(mockProductsRepo.getOneBy).toHaveBeenCalledWith({ title: 'testing' });
    });
  });

  describe('when: the db does not contain the product AND the price is a valid float', () => {
    let rsp;
    let id = '1221';

    beforeEach(async () => {
      mockProductsRepo.getOneBy = jest.fn().mockImplementation(() => {
        return new Promise((resolve) => {
          resolve(undefined);
        });
      });

      mockProductsRepo.create = jest.fn().mockImplementation(() => {
        return new Promise((resolve) => {
          resolve({ id: id });
        });
      });

      rsp = await request(app)
        .post('/admin/products/new')
        .send({ title: 'testing', price: '21.21' });
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    test("then: we return 'Submitted!'", async () => {
      const status = rsp.status;
      expect(status).toBe(200);

      const text = rsp.text;
      expect(text).toMatchSnapshot();

      expect(mockProductsRepo.getOneBy).toHaveBeenCalledWith({ title: 'testing' });
    });
  });
});
