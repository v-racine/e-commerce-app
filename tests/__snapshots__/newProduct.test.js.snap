// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create new product when: the db already contains that product then: we return 'Product already exists' 1`] = `
"
    <!DOCTYPE html>
      <html lang="en">
      <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>Shop</title>
        <link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/all.min.css" rel="stylesheet">
        <link href="/css/main.css" rel="stylesheet">
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/bulma/0.7.5/css/bulma.min.css"></link>
      </head>

      <body class="admin">
        <header>
          <nav class="navbar navbar-bottom">
            <div class="container navbar-container">
              <div>
                <a href="/admin/products">
                  <h3 class="title">Admin Panel</h3>
                </a>
              </div>
              <div class="navbar-item">
                <div class="navbar-buttons">
                  <div class="navbar-item">
                    <a href="/admin/products"><i class="fa fa-star"></i> Products</a>
                  </div>
                </div>
              </div>
            </div>
          </nav>
        </header>
        <div class="container">
          
      <div class="columns is-centered">
        <div class="column is-half">
          <h1 class="subtitle">Create a Product</h1>

          <form method="POST" enctype="multipart/form-data">
            <div class="field">
              <label class="label">Title</label>
              <input class="input" placeholder="Title" name="title">
              <p class="help is-danger"></p>
            </div>
            
            <div class="field">
              <label class="label">Price</label>
              <input class="input" placeholder="Price" name="price">
              <p class="help is-danger"></p>
            </div>
            
            <div class="field">
              <label class="label">Image</label>            
              <input type="file" name="image" />
              <p class="help is-danger"></p>
            </div>
            <br />
            <button class="button is-primary">Create</button>
            Product already exists
          </form>
        </div>
      </div>
    
        </div>
      </body>
    </html>
  "
`;

exports[`create new product when: the db does not contain the product AND the price is a valid float then: we return 'Submitted!' 1`] = `"Submitted 1221"`;

exports[`create new product when: the user passes in a title and a valid price but no image then: we return "Please upload an image" 1`] = `
"
    <!DOCTYPE html>
      <html lang="en">
      <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>Shop</title>
        <link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/all.min.css" rel="stylesheet">
        <link href="/css/main.css" rel="stylesheet">
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/bulma/0.7.5/css/bulma.min.css"></link>
      </head>

      <body class="admin">
        <header>
          <nav class="navbar navbar-bottom">
            <div class="container navbar-container">
              <div>
                <a href="/admin/products">
                  <h3 class="title">Admin Panel</h3>
                </a>
              </div>
              <div class="navbar-item">
                <div class="navbar-buttons">
                  <div class="navbar-item">
                    <a href="/admin/products"><i class="fa fa-star"></i> Products</a>
                  </div>
                </div>
              </div>
            </div>
          </nav>
        </header>
        <div class="container">
          
      <div class="columns is-centered">
        <div class="column is-half">
          <h1 class="subtitle">Create a Product</h1>

          <form method="POST" enctype="multipart/form-data">
            <div class="field">
              <label class="label">Title</label>
              <input class="input" placeholder="Title" name="title">
              <p class="help is-danger"></p>
            </div>
            
            <div class="field">
              <label class="label">Price</label>
              <input class="input" placeholder="Price" name="price">
              <p class="help is-danger"></p>
            </div>
            
            <div class="field">
              <label class="label">Image</label>            
              <input type="file" name="image" />
              <p class="help is-danger">Please upload an image</p>
            </div>
            <br />
            <button class="button is-primary">Create</button>
            
          </form>
        </div>
      </div>
    
        </div>
      </body>
    </html>
  "
`;

exports[`create new product when: the user passes in something that is not a float number then: we return 'Invalid value' 1`] = `
"
    <!DOCTYPE html>
      <html lang="en">
      <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>Shop</title>
        <link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/all.min.css" rel="stylesheet">
        <link href="/css/main.css" rel="stylesheet">
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/bulma/0.7.5/css/bulma.min.css"></link>
      </head>

      <body class="admin">
        <header>
          <nav class="navbar navbar-bottom">
            <div class="container navbar-container">
              <div>
                <a href="/admin/products">
                  <h3 class="title">Admin Panel</h3>
                </a>
              </div>
              <div class="navbar-item">
                <div class="navbar-buttons">
                  <div class="navbar-item">
                    <a href="/admin/products"><i class="fa fa-star"></i> Products</a>
                  </div>
                </div>
              </div>
            </div>
          </nav>
        </header>
        <div class="container">
          
      <div class="columns is-centered">
        <div class="column is-half">
          <h1 class="subtitle">Create a Product</h1>

          <form method="POST" enctype="multipart/form-data">
            <div class="field">
              <label class="label">Title</label>
              <input class="input" placeholder="Title" name="title">
              <p class="help is-danger"></p>
            </div>
            
            <div class="field">
              <label class="label">Price</label>
              <input class="input" placeholder="Price" name="price">
              <p class="help is-danger">Invalid value</p>
            </div>
            
            <div class="field">
              <label class="label">Image</label>            
              <input type="file" name="image" />
              <p class="help is-danger">Please upload an image</p>
            </div>
            <br />
            <button class="button is-primary">Create</button>
            
          </form>
        </div>
      </div>
    
        </div>
      </body>
    </html>
  "
`;

exports[`create new product when: the user passes in something that is not a product and no price then: we return both error messages 1`] = `
"
    <!DOCTYPE html>
      <html lang="en">
      <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>Shop</title>
        <link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/all.min.css" rel="stylesheet">
        <link href="/css/main.css" rel="stylesheet">
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/bulma/0.7.5/css/bulma.min.css"></link>
      </head>

      <body class="admin">
        <header>
          <nav class="navbar navbar-bottom">
            <div class="container navbar-container">
              <div>
                <a href="/admin/products">
                  <h3 class="title">Admin Panel</h3>
                </a>
              </div>
              <div class="navbar-item">
                <div class="navbar-buttons">
                  <div class="navbar-item">
                    <a href="/admin/products"><i class="fa fa-star"></i> Products</a>
                  </div>
                </div>
              </div>
            </div>
          </nav>
        </header>
        <div class="container">
          
      <div class="columns is-centered">
        <div class="column is-half">
          <h1 class="subtitle">Create a Product</h1>

          <form method="POST" enctype="multipart/form-data">
            <div class="field">
              <label class="label">Title</label>
              <input class="input" placeholder="Title" name="title">
              <p class="help is-danger">Must be between 5 and 40 characters</p>
            </div>
            
            <div class="field">
              <label class="label">Price</label>
              <input class="input" placeholder="Price" name="price">
              <p class="help is-danger">Invalid value</p>
            </div>
            
            <div class="field">
              <label class="label">Image</label>            
              <input type="file" name="image" />
              <p class="help is-danger">Please upload an image</p>
            </div>
            <br />
            <button class="button is-primary">Create</button>
            
          </form>
        </div>
      </div>
    
        </div>
      </body>
    </html>
  "
`;

exports[`create new product when: the user passes in something that is not a valid product title then: we return 'Must be between 5 and 40 characters' html 1`] = `
"
    <!DOCTYPE html>
      <html lang="en">
      <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>Shop</title>
        <link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/all.min.css" rel="stylesheet">
        <link href="/css/main.css" rel="stylesheet">
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/bulma/0.7.5/css/bulma.min.css"></link>
      </head>

      <body class="admin">
        <header>
          <nav class="navbar navbar-bottom">
            <div class="container navbar-container">
              <div>
                <a href="/admin/products">
                  <h3 class="title">Admin Panel</h3>
                </a>
              </div>
              <div class="navbar-item">
                <div class="navbar-buttons">
                  <div class="navbar-item">
                    <a href="/admin/products"><i class="fa fa-star"></i> Products</a>
                  </div>
                </div>
              </div>
            </div>
          </nav>
        </header>
        <div class="container">
          
      <div class="columns is-centered">
        <div class="column is-half">
          <h1 class="subtitle">Create a Product</h1>

          <form method="POST" enctype="multipart/form-data">
            <div class="field">
              <label class="label">Title</label>
              <input class="input" placeholder="Title" name="title">
              <p class="help is-danger">Must be between 5 and 40 characters</p>
            </div>
            
            <div class="field">
              <label class="label">Price</label>
              <input class="input" placeholder="Price" name="price">
              <p class="help is-danger"></p>
            </div>
            
            <div class="field">
              <label class="label">Image</label>            
              <input type="file" name="image" />
              <p class="help is-danger">Please upload an image</p>
            </div>
            <br />
            <button class="button is-primary">Create</button>
            
          </form>
        </div>
      </div>
    
        </div>
      </body>
    </html>
  "
`;
